### PR TITLE
Add math-parser library

### PR DESCRIPTION
**Pull request by urisinger**
Resolves #2026

For more info check out the refenced issue.

### CODE REVIEW

File: libraries/math-parser/src/constants.rs [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+			"cot" => Some(Value::Number(Number::Complex(complex.tan().recip()))),
+
+			"invsin" => Some(Value::Number(Number::Complex(complex.asin()))),
+			"invcos" => Some(Value::Number(Number::Complex(complex.acos()))),
+			"invtan" => Some(Value::Number(Number::Complex(complex.atan()))),
+			"invcsc" => Some(Value::Number(Number::Complex(complex.recip().asin()))),
+			"invsec" => Some(Value::Number(Number::Complex(complex.recip().acos()))),
+			"invcot" => Some(Value::Number(Number::Complex((Complex::new(PI / 2.0, 0.0) - complex).atan()))),
+
+			_ => None, // Handle unknown function names
```
Comment (0HyperCube) [inline libraries/math-parser/src/constants.rs:46]: A warning here (just to the console) might be useful.

File: libraries/math-parser/src/context.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+		self.0.insert(name, value);
+	}
+
+	pub fn contains(&self, name: &str) -> bool {
+		self.0.contains_key(name)
+	}
+
+	pub fn remove(&mut self, name: &str) -> Option<Value> {
+		self.0.remove(name)
+	}
```
Comment (0HyperCube) [inline libraries/math-parser/src/context.rs:51]: You can just implement deref mut so all of the HashMap functions are exposed.

**Review by 0HyperCube — APPROVED** [review]
Looks good; I'm sorry that it has taken so long to merge anything related to this.

**Comment by Keavon** [comment]
It looks like I don't have permission to push to your branch. Can you delete the test at the very bottom of `editor_api.rs` which is failing CI?

**Comment by Keavon** [comment]
!build

**Comment by Keavon** [comment]
From my tests, this seems to work well when replacing meval! Thank you. And sorry it took so long to get to.

File: libraries/math-parser/src/constants.rs (hunk 1 above)
Comment (urisinger) [reply libraries/math-parser/src/constants.rs:46]: when a function lookup fails we should check if an ident with that name exists, for now i changed it to be a hashmap lookup instad.

**Comment by urisinger** [comment]
> From my tests, this seems to work well when replacing meval! Thank you. And sorry it took so long to get to.

no problem, this still needs quite a bit of work to work as its own node, mainly becuase it needs some sort of compile time node generation

**Comment by Keavon** [comment]
That'll be a good step for the next PR. I'm looking forward to that! This one is good to merge now.

**Review by Keavon — APPROVED** [review]

**Pull request merged**